File: packages/gavio-js/src/interceptors/pii/scanners/ssn.ts
/** US Social Security Number scanner. */

import type { ScanContext } from '../context.js'
import { makeMatch } from '../match.js'
import type { PiiMatch } from '../match.js'
import type { PiiScanner } from '../scanner.js'

// AAA-GG-SSSS with hyphens or spaces. Requires a separator to avoid colliding
// with bare 9-digit numbers (handled by the BSN scanner / others).
const SSN = /\b(?!000|666|9\d\d)\d{3}[ -](?!00)\d{2}[ -](?!0000)\d{4}\b/g

export function ssnScanner(): PiiScanner {
  return {
    entityType: 'SSN',
    tier: 1,
    scan(text: string, ctx: ScanContext): PiiMatch[] {
      const out: PiiMatch[] = []
      for (const m of text.matchAll(SSN)) {
        const idx = ctx.nextIndex('SSN')
        out.push(
          makeMatch({
            entityType: 'SSN',
            start: m.index,
            end: m.index + m[0].length,
            value: m[0],
            replacement: `[SSN_${idx}]`,
          }),
        )
      }
      return out
    },
  }
}

File: packages/gavio-py/gavio/interceptors/pii/scanners/ssn.py
"""US Social Security Number scanner."""

from __future__ import annotations

import re

from ..context import ScanContext
from ..match import PiiMatch
from ..scanner import PiiScanner

# AAA-GG-SSSS with hyphens or spaces. Requires a separator to avoid colliding
# with bare 9-digit numbers (handled by BsnScanner / others).
_SSN = re.compile(r"\b(?!000|666|9\d\d)\d{3}[ -](?!00)\d{2}[ -](?!0000)\d{4}\b")


class SsnScanner(PiiScanner):
    entity_type = "SSN"

    def scan(self, text: str, ctx: ScanContext) -> list[PiiMatch]:
        out: list[PiiMatch] = []
        for m in _SSN.finditer(text):
            idx = ctx.next_index(self.entity_type)
            out.append(
                PiiMatch(
                    entity_type=self.entity_type,
                    start=m.start(),
                    end=m.end(),
                    value=m.group(),
                    replacement=f"[SSN_{idx}]",
                )
            )
        return out
